fuzz: Faster utxo_snapshot fuzz target

maflcko opened the pull request:
Two commits to speed up unit and fuzz tests.

Can be tested by running the fuzz target and looking at the time it took, or by looking at the flamegraph. For example:

```
FUZZ=utxo_snapshot perf record -g --call-graph dwarf ./src/test/fuzz/fuzz  -runs=100
hotspot ./perf.data

Commit:
test: Disallow fee_estimator construction in ChainTestingSetup

It is expensive to construct, and only one test uses it.

Fix both issues by disallowing the construction and moving it to the
single test that uses it.

dergoegge commented:
What is the expected speed up?

Commit:
fuzz: Speed up utxo_snapshot by lazy re-init

The re-init is expensive, so skip it if there is no need.

Also, add an even faster fuzz target utxo_snapshot_invalid, which does
not need any re-init at all.

sedited reviewed (approved):
Nice, ACK fa48ded21168740640ff9d6e5b413a322d58b114

I'd re-ack quickly if you want to fix the other includes as well.

marcofleon reviewed (commented):
Tested ACK fa198d1a196482bf363ba57a3f4d6d2d860b8ce5

marcofleon commented:
Sounds good, I'll review again when it's improved then. Looked to be about 2x when I just tested it.

Commit:
fuzz: Remove unused DataStream object

sedited reviewed (commented):
I still see a doubling in performance of the valid test if the validation cache is initialized with minimal values. Do you want to add it back?

current HEAD:
```
#10000	DONE   cov: 2794 ft: 3029 corp: 66/642b lim: 33 exec/s: 161 rss: 406Mb
Done 10000 runs in 62 second(s)
```
`+` min validation cache:
```
#10000	DONE   cov: 2788 ft: 3011 corp: 54/431b lim: 68 exec/s: 333 rss: 350Mb
Done 10000 runs in 30 second(s)
```

Commit:
fuzz: Speed up utxo_snapshot fuzz target

This speeds up the fuzz target, which allows "valid" inputs. It does not
affect the "INVALID" fuzz target.

TheCharlatan commented on src/test/fuzz/utxo_snapshot.cpp:
```diff
@@ -1,45 +1,79 @@
+#include <memory>
+#include <optional>
+#include <ranges>
+#include <vector>
```
Nit: The includes need to be corrected again.

sedited reviewed (approved):
ACK fa69afd51879f8441df7a51144318a49e7dd27b9

The logic is a bit convoluted, but I don't have many suggestions for improving it and I think it is important to get this performance improvement deployed quickly.

marcofleon commented:
![exec:s](https://github.com/user-attachments/assets/4a2bb3fa-c1e9-4238-9201-8183344006b0)
![exec:s_dos](https://github.com/user-attachments/assets/61c0b87f-188a-43cf-93ef-d0c427d833d4)

The second one just shows the same plot zoomed in on the last 600k iterations or so. I'm going to follow up with a plot of executions vs time as well. And I'll do more than a million iterations, as it seems like `utxo_snapshot_invalid` might end up being faster if fuzzed for hours.

I compiled using  only `--with-sanitizers=fuzzer`. Should I test for speed using other sanitizers or will the performance difference be similar regardless?

marcofleon commented:
Thank you for the suggestions, I'll do more trials for sure. That actually is the `utxo_snapshot_invalid` target compared to the old `utxo_snapshot` (sorry, file names should've been clearer). It's much faster at the beginning. I ran it from scratch, not on any corpus. I'm generating a corpus now for the invalid target and then I'll do a comparison for both when fuzzing from a corpus. 

Also, for the bug from https://github.com/bitcoin/bitcoin/issues/30514#issuecomment-2285605841. What's the reason for `max_len=128`? Is that necessary to find the bug in a reasonable amount of time?

marcofleon commented:
To clarify, I reproduced the bug fairly quickly with the line you provided in https://github.com/bitcoin/bitcoin/pull/30644#issuecomment-2290858777. I would like to see what happens with `max_len` not specified.

marcofleon commented:
I'll recheck my setup and get back to you.

sedited reviewed (approved):
Re-ACK fa0b66400c12d11579b38b1ac76f677b23e8c82a

I saw the old "libFuzzer disabled leak detection" info message in the invalid test, but not sure what might be causing it.

marcofleon commented:
I'm using Debian. Here's my recent run (2 million iterations) using an initial corpus for both targets:

![Screenshot 2024-08-19 at 10 54 46 AM](https://github.com/user-attachments/assets/8011c79b-ea03-4130-a946-ce52fb202655)

The new target is clearly faster. I think this is a more useful run because we generally fuzz from a corpus of inputs. I also reproduced the bug using only `-use_value_profile=1`. So overall, I would say this PR is good to go. I agree that it's probably not really worth spending time to improve it for non-ramdisk fuzzing, but up to you.

It is interesting that the new target eventually slows down significantly when fuzzing from no inputs. I ran this:
`FUZZ=utxo_snapshot_invalid perf record -g --call-graph dwarf ./src/test/fuzz/fuzz  -runs=200000`
and here is flamegraph: https://marcofleon.github.io/flamegraph/19-8-utxo_snap_invalid.svg

marcofleon commented:
Let me know if you think there is anything more I can do to test. Looking at that flame graph is fun so I'm happy to keep fuzzing.

marcofleon commented:
Makes sense, was just working with Niklas to get this done. The flame graphs are a useful new tool for me too, I'll be utilizing those more. Thanks for the help @maflcko.

marcofleon commented:
ACK fa0b66400c12d11579b38b1ac76f677b23e8c82a

TheCharlatan commented on src/test/util/setup_common.h:
```diff
@@ -28,7 +28,6 @@ class arith_uint256;
 class CFeeRate;
 class Chainstate;
 class FastRandomContext;
-
```
I don't think this empty line should be removed.

sedited reviewed (approved):
Re-ACK fa899fb7aa8a14acecadd8936ad5824fa0f697ff

marcofleon reviewed (approved):
Re ACK fa899fb7aa8a14acecadd8936ad5824fa0f697ff

brunoerg reviewed (approved):
ACK fa899fb7aa8a14acecadd8936ad5824fa0f697ff

Left it running for a while, got a clear improvement compared to master.
